### ISSUE TITLE
ClientAuthenticationError: DefaultAzureCredential failed to retrieve a token from the included credentials.

### ISSUE THREAD

**Issue by vivek-dani** [Azure/azure-sdk-for-python#29040]
- **Package Name**:  azure.identity
- **Package Version**:  1.12.0
- **Operating System**: Windows
- **Python Version**:  3.7.12

**Describe the bug**
I load ml_client as below, from config.json in local folder.   
```
credential = DefaultAzureCredential()
ml_client = MLClient(credential, subscription_id, resource_group, workspace_name)
```
I get below error when ml_client trys to get the token internally. Its collecting stale token from cache.

```
File "C:\go\setup\anaconda\envs\e2e_test\lib\site-packages\azure\identity\_credentials\chained.py", line 108, in get_token
    raise ClientAuthenticationError(message=message)
azure.core.exceptions.ClientAuthenticationError: DefaultAzureCredential failed to retrieve a token from the included credentials.
Attempted credentials:
        EnvironmentCredential: EnvironmentCredential authentication unavailable. Environment variables are not fully configured.
Visit https://aka.ms/azsdk/python/identity/environmentcredential/troubleshoot to troubleshoot.this issue.
        ManagedIdentityCredential: ManagedIdentityCredential authentication unavailable, no response from the IMDS endpoint.
        SharedTokenCacheCredential: Azure Active Directory error '(invalid_grant) AADSTS50173: The provided grant has expired due to it being revoked, a fresh auth token is needed. The user might have changed or reset their password. The grant was issued on '2023-02-09T17:58:24.8491242Z' and the TokensValidFrom date (before which tokens are not valid) for this user is '2023-02-18T15:27:11.0000000Z'.
Trace ID: 6849addd-100a-497e-887e-c2bc6f5c0300
Correlation ID: a1dc633f-0908-4613-8248-ad10ee704be8
Timestamp: 2023-02-21 09:34:07Z'
Content: {"error":"invalid_grant","error_description":"AADSTS50173: The provided grant has expired due to it being revoked, a fresh auth token is needed. The user might have changed or reset their password. The grant was issued on '2023-02-09T17:58:24.8491242Z' and the TokensValidFrom date (before which tokens are not valid) for this user is '2023-02-18T15:27:11.0000000Z'.\r\nTrace ID: 6849addd-100a-497e-887e-c2bc6f5c0300\r\nCorrelation ID: a1dc633f-0908-4613-8248-ad10ee704be8\r\nTimestamp: 2023-02-21 09:34:07Z","error_codes":[50173],"timestamp":"2023-02-21 09:34:07Z","trace_id":"6849addd-100a-497e-887e-c2bc6f5c0300","correlation_id":"a1dc633f-0908-4613-8248-ad10ee704be8","error_uri":"https://login.microsoftonline.com/error?code=50173"}
To mitigate this issue, please refer to the troubleshooting guidelines here at https://aka.ms/azsdk/python/identity/defaultazurecredential/troubleshoot.
```

- I am able to authenticate using InteractiveBrowserCredential()
- I am able to authenticate using cliI don't remember changing password recently
- I tried, to do az upgrade, az login, clear cache in .azure folder under C://Users/<alias>; but DefaultAzureCredential() didn't work. It was still taking stale token.

I was able to resolve above issue by manually deleting msal.cache , msalV2.cache files under .IdentityService folder, under os.environ["LOCALAPPDATA"]

**Comment by xiangyan99** [comment]
It looks to me that you have logged in using Visual Studio and the token has expired.

Can you try 

```py
credential = DefaultAzureCredential(exclude_shared_token_cache_credential=True)
ml_client = MLClient(credential, subscription_id, resource_group, workspace_name)
```

to see if it works?

**Comment by vishveshtrivedi** [comment]
Hello,
I am having Similar Issue.
When I try to run the following thing on gitlab it gives me error.
if credential is None:
         credential = azure.identity.DefaultAzureCredential(exclude_shared_token_cache_credential=True)
        self.credential = credential
        if kv_name is None:
            kv_name = os.getenv("AZURE_KV_NAME")
        url = f"https://{kv_name}.vault.azure.net"
        super().__init__(url, credential, **kwargs)

Result: Failure
                                                                        Exception: ClientAuthenticationError: DefaultAzureCredential failed to retrieve a token from the included credentials.
                                                                        Attempted credentials:
                                                                        	EnvironmentCredential: EnvironmentCredential authentication unavailable. Environment variables are not fully configured.
                                                                        Visit https://aka.ms/azsdk/python/identity/environmentcredential/troubleshoot to troubleshoot.this issue.
                                                                        	ManagedIdentityCredential: ManagedIdentityCredential authentication unavailable, no response from the IMDS endpoint.
                                                                        	SharedTokenCacheCredential: SharedTokenCacheCredential authentication unavailable. No accounts were found in the cache.
                                                                        	AzureCliCredential: Azure CLI not found on path
                                                                        	AzurePowerShellCredential: PowerShell is not installed
                                                                        To mitigate this issue, please refer to the troubleshooting guidelines here at https://aka.ms/azsdk/python/identity/defaultazurecredential/troubleshoot.
                                                                        Stack:   File "/azure-functions-host/workers/python/3.9/LINUX/X64/azure_functions_worker/dispatcher.py", line 452, in _handle__invocation_request
                                                                            call_result = await self._loop.run_in_executor(
                                                                          File "/usr/local/lib/python3.9/concurrent/futures/thread.py", line 58, in run
                                                                            result = self.fn(*self.args, **self.kwargs)
                                                                          File "/azure-functions-host/workers/python/3.9/LINUX/X64/azure_functions_worker/dispatcher.py", line 718, in _run_sync_func
                                                                            return ExtensionManager.get_sync_invocation_wrapper(context,
                                                                          File "/azure-functions-host/workers/python/3.9/LINUX/X64/azure_functions_worker/extension.py", line 215, in _raw_invocation_wrapper
                                                                            result = function(**args)
                                                                          File "/home/site/wwwroot/regatta_association/__init__.py", line 26, in main
                                                                            con_in = api.get_northstar_connection(secrets_manager := api.SecretsManager())
                                                                          File "/home/site/wwwroot/api/database.py", line 43, in get_northstar_connection
                                                                            username = secrets_manager.get_secret_value("northstar-db-username")
                                                                          File "/home/site/wwwroot/api/secrets.py", line 39, in get_secret_value
                                                                            return self.get_secret(name, version, **kwargs).value  # type: ignore
                                                                          File "/home/site/wwwroot/.python_packages/lib/site-packages/azure/core/tracing/decorator.py", line 78, in wrapper_use_tracer
                                                                            return func(*args, **kwargs)
                                                                          File "/home/site/wwwroot/.python_packages/lib/site-packages/azure/keyvault/secrets/_client.py", line 72, in get_secret
                                                                            bundle = self._client.get_secret(
                                                                          File "/home/site/wwwroot/.python_packages/lib/site-packages/azure/keyvault/secrets/_generated/_operations_mixin.py", line 1574, in get_secret
                                                                            return mixin_instance.get_secret(vault_base_url, secret_name, secret_version, **kwargs)
                                                                          File "/home/site/wwwroot/.python_packages/lib/site-packages/azure/core/tracing/decorator.py", line 78, in wrapper_use_tracer
                                                                            return func(*args, **kwargs)
                                                                          File "/home/site/wwwroot/.python_packages/lib/site-packages/azure/keyvault/secrets/_generated/v7_3/operations/_key_vault_client_operations.py", line 694, in get_secret
                                                                            pipeline_response = self._client._pipeline.run(  # pylint: disable=protected-access
                                                                          File "/home/site/wwwroot/.python_packages/lib/site-packages/azure/core/pipeline/_base.py", line 205, in run
                                                                            return first_node.send(pipeline_request)  # type: ignore
                                                                          File "/home/site/wwwroot/.python_packages/lib/site-packages/azure/core/pipeline/_base.py", line 69, in send
                                                                            response = self.next.send(request)
                                                                          File "/home/site/wwwroot/.python_packages/lib/site-packages/azure/core/pipeline/_base.py", line 69, in send
                                                                            response = self.next.send(request)
                                                                          File "/home/site/wwwroot/.python_packages/lib/site-packages/azure/core/pipeline/_base.py", line 69, in send
                                                                            response = self.next.send(request)
                                                                          [Previous line repeated 2 more times]
                                                                          File "/home/site/wwwroot/.python_packages/lib/site-packages/azure/core/pipeline/policies/_redirect.py", line 160, in send
                                                                            response = self.next.send(request)
                                                                          File "/home/site/wwwroot/.python_packages/lib/site-packages/azure/core/pipeline/policies/_retry.py", line 474, in send
                                                                            response = self.next.send(request)
                                                                          File "/home/site/wwwroot/.python_packages/lib/site-packages/azure/core/pipeline/policies/_authentication.py", line 126, in send
                                                                            request_authorized = self.on_challenge(request, response)
                                                                          File "/home/site/wwwroot/.python_packages/lib/site-packages/azure/keyvault/secrets/_shared/challenge_auth_policy.py", line 117, in on_challenge
                                                                            self.authorize_request(request, scope, tenant_id=challenge.tenant_id)
                                                                          File "/home/site/wwwroot/.python_packages/lib/site-packages/azure/core/pipeline/policies/_authentication.py", line 106, in authorize_request
                                                                            self._token = self._credential.get_token(*scopes, **kwargs)
                                                                          File "/home/site/wwwroot/.python_packages/lib/site-packages/azure/identity/_credentials/default.py", line 168, in get_token
                                                                            return super(DefaultAzureCredential, self).get_token(*scopes, **kwargs)
                                                                          File "/home/site/wwwroot/.python_packages/lib/site-packages/azure/identity/_credentials/chained.py", line 101, in get_token
                                                                            raise ClientAuthenticationError(message=message)


Can anyone help me out in this?

**Comment by xiangyan99** [comment]
@vishveshtrivedi from your log, seems like there was no valid credential found.

DefaultAzureCredential is a chain of credentials. It attempts different credentials in order until it finds one. Or it raise the ClientAuthenticationError.

Which credential did you expect to work?

**Comment by vishveshtrivedi** [comment]
Actually the code was running properly till yesterday.
It will automatically take credential method. We do not provide any arguments to the same.
For the local testing we were using interactive browser credential.
for the local connection we are using following piece of code:

api.get_northstar_connection(secrets_manager := api.SecretsManager(credential = InteractiveBrowserCredential(additionally_allowed_tenants=['*'])))

But when we deploy to gitlab we are using following:
con_in = api.get_northstar_connection(secrets_manager := api.SecretsManager())

**Comment by xiangyan99** [comment]
I am not an expert of gitlab.

Is it possible that the token got from InteractiveBrowserCredential expired?

You may need to log in again.

**Comment by vishveshtrivedi** [comment]
I logged in again but the same error is there.

**Comment by xiangyan99** [comment]
Still with same error log?

I wanted to double check because your log showed there was no valid login information found.

If you logged in, it should be able to find one?

**Comment by vishveshtrivedi** [comment]
Actually the thing is I logged in when I run the code locally using interactive browser credential.

But when we upload to gitlab we make to use default credential.

If possible can we do a microsoft teams call so that I can show the exact problem?

**Comment by vivek-dani** [comment]
> It looks to me that you have logged in using Visual Studio and the token has expired.
> 
> Can you try
> 
> ```python
> credential = DefaultAzureCredential(exclude_shared_token_cache_credential=True)
> ml_client = MLClient(credential, subscription_id, resource_group, workspace_name)
> ```
> 
> to see if it works?

No I have never used VisualStudio.
DefaultAzureCredential(exclude_shared_token_cache_credential=True) would work for normal tasks, but for MlFlowClient object as used here https://github.com/Azure/azureml-examples/blob/main/sdk/python/jobs/automl-standalone-jobs/automl-image-classification-multiclass-task-fridge-items/automl-image-classification-multiclass-task-fridge-items.ipynb  DefaultAzureCredential(exclude_shared_token_cache_credential=True) won't work

**Comment by vivek-dani** [comment]
Problem is its always taking the stale/old credentials.
There are msal.cache , msalV2.cache files under .IdentityService folder, under os.environ["LOCALAPPDATA"]. And I had to explicitly delete these files manually to get unblocked

**Comment by vishveshtrivedi** [comment]
I was able to login using managedidentity by adding my function app in keyvault as my functin was reading secrets from key vault.
I have another issue with database and following is the error:

Exception: InterfaceError: (pyodbc.InterfaceError) ('28000', "[28000] [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]Login failed for user '<token-identified principal>'. (18456) (SQLDriverConnect)")

Does anyone know how to fix this?

**Comment by xiangyan99** [comment]
> > It looks to me that you have logged in using Visual Studio and the token has expired.
> > Can you try
> > ```python
> > credential = DefaultAzureCredential(exclude_shared_token_cache_credential=True)
> > ml_client = MLClient(credential, subscription_id, resource_group, workspace_name)
> > ```
> >   
> > to see if it works?
> 
> @xiangyan99 xiangyan99 No I have never used VisualStudio for my work. DefaultAzureCredential(exclude_shared_token_cache_credential=True) would work for normal tasks, but for MlFlowClient object as used here https://github.com/Azure/azureml-examples/blob/main/sdk/python/jobs/automl-standalone-jobs/automl-image-classification-multiclass-task-fridge-items/automl-image-classification-multiclass-task-fridge-items.ipynb DefaultAzureCredential(exclude_shared_token_cache_credential=True) won't work

Could you give more details about 

but for MlFlowClient object, DefaultAzureCredential(exclude_shared_token_cache_credential=True) won't work?

exclude_shared_token_cache_credential=True should make DefaultAzureCredential opt out to use cached token.

**Comment by xiangyan99** [comment]
@vishveshtrivedi Looks like the failure has different causes.

Please open a separate issue to track it.

**Comment by vishveshtrivedi** [comment]
Sure.

**Comment by vishveshtrivedi** [comment]
Issue has been solved by adding managed identity.
You can close the issue.
Thanks for the help.

**Comment by vivek-dani** [comment]
@xiangyan99 Bug defined in this issue is different from what @vishveshtrivedi is facing. Problem faced by @vishveshtrivedi is solved, however the bug defined in this issue still exists. Requesting to resolve this issue.

**Comment by xiangyan99** [comment]
Took offline and waiting for author feedback.
